### ISSUE TITLE
asm2wasm: local var initialization silently dropped

### ISSUE THREAD

**Issue by mnater** [WebAssembly/binaryen#1402]
asm2wasm compiles this (handwritten and afaik valid) asm.js function
````javascript
function add10fails(i) {
    i = i | 0;
    var k = 10;
    return (i + k) | 0;
}
````
to
````
(func $add10fails (; 0 ;) (param $i i32) (result i32)
  (local $k i32)
  (return
   (i32.add
    (get_local $i)
    (get_local $k)
   )
  )
 )
````
Thus, the initialization value for the local variable `k` is silently dropped. It works if I initialize `k` in a separate statement:
````javascript
function add10works(i) {
    i = i | 0;
    var k = 0;
    k = 10 | 0;
    return (i + k) | 0;
}
````
which correctly compiles to
````
 (func $add10works (; 1 ;) (param $i i32) (result i32)
  (local $k i32)
  (set_local $k
   (i32.const 10)
  )
  (return
   (i32.add
    (get_local $i)
    (get_local $k)
   )
  )
 )
````

I know asm2wasm is meant for emscripten output. But I think asm2wasm should at least emit a warning here.

**Comment by kripken** [comment]
The issue here is that `var k = 10;` is not valid asm.js, and asm2wasm assumes the input is valid.

I agree we should add error messages for code that isn't asm.js, but our parser isn't currently designed in a way that makes that simple. It would be nice though if someone were interested to write a better one, or improve the existing one.

**Comment by mnater** [comment]
> The issue here is that var k = 10; is not valid asm.js, and asm2wasm assumes the input is valid.

In my understanding of http://asmjs.org/spec/latest/#variable-type-annotations `var k = 10;` was valid asm.js. At least it passes all asm.js validators I tried with no error...
Could you explain in Layman's Terms why this isn't valid?

**Comment by kripken** [comment]
Good point, looks like the spec does allow it. I think that changed at some point, as in early versions it mandated only 0 in the declarations.

In that case, it would be nice to support it in asm2wasm. I'm not sure offhand how easy that would be, but I think there are some hardcoded assumptions on 0 in declarations.
